xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorViewController.swift

```diff
@@ -0,0 +1,1211 @@
+        model.blocks[i] = transform(block)
+        return model
+    }
+
+    func selectBlock(
+        state: Model,
+        id: UUID,
+        isSelected: Bool
+    ) -> Update {
+        guard state.isBlockSelectMode else {
+            Self.logger.log("block#\(id) selected, but not in select mode. Doing nothing.")
+            return Update(state: state)
+        }
+        
+        guard let i = state.block(id: id) else {
+            return Update(state: state)
+        }
+        
+        var model = state
+        let block = model.blocks[i]
+        let updatedBlock = block.setBlockSelected(isSelected) ?? block
+
+        guard block != updatedBlock else {
+            Self.logger.debug("Block selection state did not change.")
+            return Update(state: state)
+        }
+        model.blocks[i] = updatedBlock
+        
+        let render = {
+            self.collectionView.reconfigureItems(
+                at: [
+                    IndexPath(
+                        row: i,
+                        section: BlockEditor.Section.blocks.rawValue
+                    )
+                ]
+            )
+        }
+        
+        return Update(state: model, render: render)
+    }
+
+    func toggleSelectBlock(
+        state: Model,
+        id: UUID
+    ) -> Update {
+        guard state.isBlockSelectMode else {
+            Self.logger.log("block#\(id) selected, but not in select mode. Doing nothing.")
+            return Update(state: state)
+        }
+        
+        guard let i = state.block(id: id) else {
+            return Update(state: state)
+        }
+        
+        var model = state
+        let block = model.blocks[i]
+        let isSelected = block.isBlockSelected
+        let updatedBlock = block.setBlockSelected(!isSelected) ?? block
+
+        guard block != updatedBlock else {
+            Self.logger.debug("Block selection state did not change.")
+            return Update(state: state)
+        }
+        model.blocks[i] = updatedBlock
+        
+        let render = {
+            self.collectionView.reconfigureItems(
+                at: [
+                    IndexPath(
+                        row: i,
+                        section: BlockEditor.Section.blocks.rawValue
+                    )
+                ]
+            )
+        }
+        
+        return Update(state: model, render: render)
+    }
+
+    func moveBlockUp(
+        state: Model,
+        id: UUID
+    ) -> Update {
+        guard let i = state.blocks.firstIndex(whereID: id) else {
+            Self.logger.log("block#\(id) not found. Doing nothing.")
+            return Update(state: state)
+        }
+        
+        var blocksArray = state.blocks
+        
+        guard i > blocksArray.startIndex else {
+            Self.logger.log("block#\(id) can't move up first block. Doing nothing.")
+            return Update(state: state)
+        }
+        
+        let h = blocksArray.index(before: i)
```

Comment:
Took me a shamefully long time to realise that this is because "h" comes before "i".